### PR TITLE
Improve output when fatal errors encountered

### PR DESCRIPTION
**Pull request by kyleknap**
Now if a fatal error is encountered it will print out the fatal error and stop printing out any more updates it may receive as it cleans up. It was really confusing when all of the transfers
cancelled by Cntrl-C would show up in results as failures. Here is an example of how this looks now:
```

$ aws s3 cp aws-cli s3://kyleknap-dump/aws-cli --recursive
upload: aws-cli/.changes/1.10.13.json to s3://kyleknap-dump/aws-cli/.changes/1.10.13.json
upload: aws-cli/.changes/1.10.14.json to s3://kyleknap-dump/aws-cli/.changes/1.10.14.json
upload: aws-cli/.changes/1.10.11.json to s3://kyleknap-dump/aws-cli/.changes/1.10.11.json
upload: aws-cli/.changes/1.10.10.json to s3://kyleknap-dump/aws-cli/.changes/1.10.10.json
upload: aws-cli/.changes/1.1.2.json to s3://kyleknap-dump/aws-cli/.changes/1.1.2.json
upload: aws-cli/.changes/1.1.1.json to s3://kyleknap-dump/aws-cli/.changes/1.1.1.json
upload: aws-cli/.changes/1.10.17.json to s3://kyleknap-dump/aws-cli/.changes/1.10.17.json
upload: aws-cli/.changes/1.10.18.json to s3://kyleknap-dump/aws-cli/.changes/1.10.18.json
upload: aws-cli/.changes/1.10.2.json to s3://kyleknap-dump/aws-cli/.changes/1.10.2.json
upload: aws-cli/.changes/1.10.19.json to s3://kyleknap-dump/aws-cli/.changes/1.10.19.json
upload: aws-cli/.changes/1.10.16.json to s3://kyleknap-dump/aws-cli/.changes/1.10.16.json
upload: aws-cli/.changes/1.10.20.json to s3://kyleknap-dump/aws-cli/.changes/1.10.20.json
upload: aws-cli/.changes/1.10.15.json to s3://kyleknap-dump/aws-cli/.changes/1.10.15.json
upload: aws-cli/.changes/1.1.0.json to s3://kyleknap-dump/aws-cli/.changes/1.1.0.json
upload: aws-cli/.changes/1.10.12.json to s3://kyleknap-dump/aws-cli/.changes/1.10.12.json
fatal error: keyboard interrupt                       
$ 
```
I decided not to include the cleaning up bit because now Cntrl-C's are really fast and is not needed plus I think it looks a little better.

Note that this PR builds off of this one: https://github.com/boto/s3transfer/pull/56. So tests may fail until the linked PR is merged.

cc @jamesls @JordonPhillips

### CODE REVIEW

**Comment by JordonPhillips** [comment]
⛵  once the dependent prs get merged / this gets rebased. This is much better than seeing a bunch of `upload failed` messages.

**Comment by kyleknap** [comment]
Everything has been merged in s3transfer so this PR and this got rebased so this should be good to look at.

**Comment by jamesls** [comment]
Code wise this looks fine, but I wonder if the terminology is confusing here.  The error messages make sense from an implementation details standpoint but I think we should show something more user friendly because:

* While it's `fatal` in the sense that it stops the transfer, it's perfectly normal to ctrl-c out of a transfer if you don't want it to keep going and don't want to convey to users that something's gone wrong
* keyboard interrupt is super specific to a python developer

What about something like "cancelled: ctrl-c received"?

**Comment by kyleknap** [comment]
@jamesls I would be fine with "cancelled: ctrl-c received" as it is more universal and human readable. 

Do you think we should be using "cancelled" for general errors that stop the entire transfer as well such as ListObjects failing?

**Comment by jamesls** [comment]
Well in this case I was thinking we want to distinguish between an "expected fatal error" where a customer has explicitly asked to shutdown (via ctrl-c) vs. a truly fatal error like the `ListObjects` failing, which is unexpected and should make it clear to the user that something unexpected went wrong.

**Comment by kyleknap** [comment]
Makes sense I will add some logic to differentiate between the two then in the result printer. Does ``"cancelled:"`` and ``"fatal error:"`` work? Or do you want something more explicit like ``"unexpected fatal error:"``? I think I like the latter more.

**Comment by kyleknap** [comment]
@jamesls I updated based on feedback so it should be good to look at.

File: awscli/customizations/s3/results.py [hunk 1]
```diff
@@ -91,7 +94,13 @@ def _on_success(self, future):
 
     def _on_failure(self, future, e):
         result_kwargs = self._on_done_pop_from_result_kwargs_cache(future)
-        self._result_queue.put(FailureResult(exception=e, **result_kwargs))
+        if isinstance(e, CancelledError):
+            error_result_cls = ErrorResult
+            if 'KeyboardInterrupt' in str(e):
```
Comment (jamesls) [inline awscli/customizations/s3/results.py]: This seems fragile, shouldn't `s3transfer` expose a better way to disambiguate between these?  What about a `UserCanelledError`?

**Comment by kyleknap** [comment]
@jamesls I just updated to rely on the new ``FatalError`` for ``ErrorResults`` instead of looking at the message of the error.

File: awscli/customizations/s3/results.py [hunk 2]
```diff
@@ -57,6 +59,8 @@ def _create_new_result_cls(name, extra_fields=None, base_cls=BaseResult):
 
 ErrorResult = namedtuple('ErrorResult', ['exception'])
 
+CntrlCResult = _create_new_result_cls('CntrlCResult', base_cls=ErrorResult)
```
Comment (jamesls) [inline awscli/customizations/s3/results.py]: Small detail, but the error messages use "Ctrl" which I think is a more common way to shorten "control".  I'd prefer to either use CtrlCResult or just the full out "ControlCResult"

File: awscli/customizations/s3/results.py [hunk 3]
```diff
@@ -310,8 +320,9 @@ class ResultPrinter(BaseResultHandler):
         '{message}'
     )
     ERROR_FORMAT = (
-        '{exception}'
+        'fatal error: {exception}'
     )
+    CNTRL_C_MSG = 'cancelled: ctrl-c received'
```
Comment (jamesls) [inline awscli/customizations/s3/results.py]: same thing, `ctrl` vs. `cntrl`.

**Comment by kyleknap** [comment]
Updated the spelling. Test pass. Merging.

**Pull request merged**